producer/src/main/kotlin/streams/StreamsTransactionEventHandler.kt

```diff
@@ -135,15 +91,94 @@ class StreamsTransactionEventHandler(private val router : StreamsEventRouter) :
         }
 
         //don't change the order of the with methods
-        val builder = PreviousTransactionDataBuilder()
-                .withLabels(txd.assignedLabels(),removedLabels)
+        return builder.withLabels(txd.assignedLabels(),removedLabels)
                 .withNodeProperties(txd.assignedNodeProperties(),removedNodeProperties)
-                .withCreatedPayloads(createdPayload)
-                .withDeletedPayloads(deletedPayload)
+                .withNodeCreatedPayloads(createdPayload)
+                .withNodeDeletedPayloads(deletedPayload)
+                .withDeletedLabels(deletedLabels)
+    }
 
-        return builder.build()
+    fun buildRelationshipChanges(txd: TransactionData, builder:PreviousTransactionDataBuilder):PreviousTransactionDataBuilder{
+        val deletedRelProperties = txd.removedRelationshipProperties().filter { txd.deletedRelationships().contains( it.entity() )}
+                .map { it -> Pair(it.entity().id, Pair(it.key(), it.previouslyCommitedValue())) }
+                .groupBy({it.first},{it.second}) // { nodeId -> [(k,v)] }
+                .mapValues { it -> it.value.toMap() }
+
+        val createdRelPayload = txd.createdRelationships().map {
+
+            val afterRel = RelationshipChangeBuilder()
+                    .withProperites(it.allProperties)
+                    .build()
+
+            val payload = RelationshipPayloadBuilder()
+                    .withId(it.id.toString())
+                    .withName(it.type.name())
+                    .withStartNode(it.startNode.id.toString(), it.startNode.labels.map { it.name() })
```

Comment:
we can add an labelNames extension method to Node

Reply:
probably also a method that takes a node and returns it's "Identification" i.e. label(s) + key-properties